File: firmware/test/ws/test_ws2.js
/*!
  Example code for node.js and the VSCP daemon ws2 websocket interface
  send 0,30,5,,,0,-,15,14,13,12,11,10,9,8,7,6,5,4,3,2,0,0,1,35
  Code from: https://github.com/websockets/ws

  Copyright 2020 Ake Hedman, the VSCP project - MIT license
*/

WebSocket = require('ws');
const crypto = require('crypto');

const defaultUrl = 'ws://localhost:8884/ws2';
const DEFAULT_USERNAME = 'admin';
const DEFAULT_PASSWORD = 'secret';
const DEFAULT_KEY_HEX = 'A4A86F7D7E119BA3F0CD06881E371B98';

function parseCliArgs() {
  const args = process.argv.slice(2);
  const parsed = {
    help: false,
    url: null,
    username: null,
    password: null,
    keyHex: null,
    positionals: []
  };

  for (let index = 0; index < args.length; index++) {
    const arg = args[index];

    if (arg === '--help' || arg === '-h') {
      parsed.help = true;
      continue;
    }

    if (arg === '--url' && args[index + 1]) {
      parsed.url = args[++index];
      continue;
    }

    if (arg.startsWith('--url=')) {
      parsed.url = arg.substring('--url='.length);
      continue;
    }

    if (arg === '--user' && args[index + 1]) {
      parsed.username = args[++index];
      continue;
    }

    if (arg.startsWith('--user=')) {
      parsed.username = arg.substring('--user='.length);
      continue;
    }

    if (arg === '--password' && args[index + 1]) {
      parsed.password = args[++index];
      continue;
    }

    if (arg.startsWith('--password=')) {
      parsed.password = arg.substring('--password='.length);
      continue;
    }

    if (arg === '--key' && args[index + 1]) {
      parsed.keyHex = args[++index];
      continue;
    }

    if (arg.startsWith('--key=')) {
      parsed.keyHex = arg.substring('--key='.length);
      continue;
    }

    if (!arg.startsWith('-')) {
      parsed.positionals.push(arg);
    }
  }

  return parsed;
}

function getAuthConfig() {
  const cli = parseCliArgs();

  if (cli.help) {
    console.log('Usage: node test_ws2.js [url] [--url <url>] [--user <username>] [--password <password>] [--key <aes128-hex>]');
    console.log('Example: node test_ws2.js --url ws://192.168.1.100:8884/ws2 --user admin --password secret --key A4A86F7D7E119BA3F0CD06881E371B98');
    console.log('Env: WS2_URL, WS2_USERNAME, WS2_PASSWORD, WS2_KEY (fallback: VSCP_USERNAME, VSCP_PASSWORD, VSCP_KEY16)');
    process.exit(0);
  }

  const wsUrl = cli.url || cli.positionals[0] || process.env.WS2_URL || defaultUrl;
  const username = cli.username || process.env.WS2_USERNAME || process.env.VSCP_USERNAME || DEFAULT_USERNAME;
  const password = cli.password || process.env.WS2_PASSWORD || process.env.VSCP_PASSWORD || DEFAULT_PASSWORD;
  const keyHex = (cli.keyHex || process.env.WS2_KEY || process.env.VSCP_KEY16 || DEFAULT_KEY_HEX).trim();

  return {
    wsUrl,
    username,
    password,
    keyHex
  };
}

function isValidHex(value, expectedLength) {
  return typeof value === 'string' && value.length === expectedLength && /^[0-9a-fA-F]+$/.test(value);
}

function encryptAuthCredentials(username, password, sidHex, keyHex) {
  if (!isValidHex(sidHex, 32)) {
    throw new Error(`Invalid SID for IV: expected 32 hex chars, got '${sidHex}'`);
  }

  if (!isValidHex(keyHex, 32)) {
    throw new Error('Invalid AES-128 key: expected 32 hex chars');
  }

  const iv = Buffer.from(sidHex, 'hex');
  const key = Buffer.from(keyHex, 'hex');
  const plainCredentials = Buffer.from(`${username};${password}`, 'utf8');

  const cipher = crypto.createCipheriv('aes-128-cbc', key, iv);
  const encrypted = Buffer.concat([cipher.update(plainCredentials), cipher.final()]);
  return encrypted.toString('hex');
}

const authConfig = getAuthConfig();
const wsUrl = authConfig.wsUrl;
console.log(`Connecting to ${wsUrl}`);

const ws = new WebSocket(wsUrl, {
  perMessageDeflate: false
});

const REQUEST_TIMEOUT_MS = 5000;
let pendingRequest = null;
let pendingSidResolver = null;

function toTextMessage(payload) {
  if (typeof payload === 'string') {
    return payload;
  }

  if (Buffer.isBuffer(payload)) {
    return payload.toString('utf8');
  }

  if (payload instanceof ArrayBuffer) {
    return Buffer.from(payload).toString('utf8');
  }

  if (ArrayBuffer.isView(payload)) {
    return Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength).toString('utf8');
  }

  return String(payload);
}

function printWs2Message(reply, rawMessage) {
  if (!reply || typeof reply !== 'object') {
    console.log("\n* * * Message received");
    console.log(rawMessage);
    return;
  }

  const msgType = reply.type || 'unknown';
  console.log(`\n* * * WS2 message (${msgType})`);

  if (msgType === 'cmd') {
    console.log(`Command   : ${reply.command || '(none)'}`);
    if (reply.result !== undefined) {
      console.log(`Result    : ${reply.result}`);
    }
    if (reply.code !== undefined) {
      console.log(`Code      : ${reply.code}`);
    }
    if (reply.message) {
      console.log(`Message   : ${reply.message}`);
    }
    if (reply.args !== undefined && reply.args !== null) {
      console.log(`Args      : ${JSON.stringify(reply.args)}`);
    }
    return;
  }

  if (msgType === 'event') {
    const event = reply.event || {};
    console.log(`Class/Type: ${event.vscpClass ?? event.class ?? '?'}/${event.vscpType ?? event.type ?? '?'}`);
    console.log(`Head/ObId : ${event.vscpHead ?? event.head ?? '?'}/${event.vscpObId ?? event.obid ?? '?'}`);
    console.log(`Date/Time : ${event.vscpDateTime ?? event.datetime ?? '?'}`);
    console.log(`Timestamp : ${event.vscpTimeStamp ?? event.timestamp ?? '?'}`);
    console.log(`GUID      : ${event.vscpGuid ?? event.guid ?? '?'}`);
    if (Array.isArray(event.vscpData || event.data)) {
      console.log(`Data      : [${(event.vscpData || event.data).join(', ')}]`);
    }
    return;
  }

  if (msgType === 'error') {
    console.log(`Code      : ${reply.code ?? '(none)'}`);
    console.log(`Message   : ${reply.message ?? '(none)'}`);
    return;
  }

  console.log(JSON.stringify(reply, null, 2));
}

function getCommandFromFrame(frame) {
  if (!frame || frame.type !== 'cmd') {
    return '';
  }

  return String(frame.command || '').toLowerCase();
}

function commandMatches(reply, expectedCommand) {
  if (!expectedCommand) {
    return true;
  }

  if (!reply || reply.type !== 'cmd') {
    return false;
  }

  return String(reply.command || '').toLowerCase() === expectedCommand.toLowerCase();
}

function findSidValue(value) {
  if (typeof value === 'string' && /^[0-9a-fA-F]{32}$/.test(value)) {
    return value;
  }

  if (Array.isArray(value)) {
    for (const item of value) {
      const sid = findSidValue(item);
      if (sid) {
        return sid;
      }
    }
  }
  else if (value && typeof value === 'object') {
    for (const objectValue of Object.values(value)) {
      const sid = findSidValue(objectValue);
      if (sid) {
        return sid;
      }
    }
  }

  return null;
}

function extractSidFromReply(reply) {
  if (!reply || typeof reply !== 'object') {
    return null;
  }

  return findSidValue(reply);
}

function waitForSid() {
  if (pendingSidResolver) {
    return Promise.reject(new Error('Already waiting for SID'));
  }

  return new Promise((resolve, reject) => {
    let challengeSent = false;
    let timeoutId;

    function armTimeout() {
      timeoutId = setTimeout(() => {
        if (!challengeSent) {
          challengeSent = true;
          console.log('* * * SID wait timed out, requesting CHALLENGE...');
          ws.send(JSON.stringify({
            type: 'cmd',
            command: 'challenge',
            args: null
          }));
          armTimeout();
          return;
        }

        pendingSidResolver = null;
        reject(new Error('Timeout waiting for SID (including CHALLENGE retry)'));
      }, REQUEST_TIMEOUT_MS);
    }

    pendingSidResolver = {
      resolve: (sid) => {
        clearTimeout(timeoutId);
        pendingSidResolver = null;
        resolve(sid);
      },
      reject: (error) => {
        clearTimeout(timeoutId);
        pendingSidResolver = null;
        reject(error);
      }
    };

    armTimeout();
  });
}

function sendAndWait(commandFrame, expectedCommand) {
  if (pendingRequest) {
    return Promise.reject(new Error('A command is already waiting for response'));
  }

  const expected = expectedCommand || getCommandFromFrame(commandFrame);

  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      pendingRequest = null;
      reject(new Error(`Timeout waiting for response to command: ${expected || '(unknown)'}`));
    }, REQUEST_TIMEOUT_MS);

    pendingRequest = {
      expected,
      resolve: (reply) => {
        clearTimeout(timeoutId);
        pendingRequest = null;
        resolve(reply);
      },
      reject: (error) => {
        clearTimeout(timeoutId);
        pendingRequest = null;
        reject(error);
      }
    };

    ws.send(JSON.stringify(commandFrame));
  });
}

ws.on('open', async function open() {
  const cmdnoop = {
    "type": "cmd",
    "command": "noop",
    "args": null
  };

  const cmdver = {
    "type": "cmd",
    "command": "version",
    "args": null
  };

  const cmdcopy = {
    "type": "cmd",
    "command": "copyright",
    "args": null
  };

  const cmdopen = {
    "type": "cmd",
    "command": "open",
    "args": null
  };

  const sendEvent = {
    "type": "event",
    "event" : {
        "vscpHead" : 0,
        "vscpObId": 0,
        "vscpDateTime": "2020-01-29T23:05:59Z",
        "vscpTimeStamp": 0,
        "vscpClass": 30,
        "vscpType": 5,
        "vscpGuid": "FF:FF:FF:FF:FF:FF:FF:F5:00:00:00:00:00:02:00:00",
        "vscpData": [1,2,3,4,5,6]
    }
  };

  // Receive only CLASS1.CONTROL, TurnOn
  const cmdFilter = {
    "type": "cmd",
    "command": "setfilter",
    "args": {                                                                  
        "mask_priority": 0,
        "mask_class": 65535,
        "mask_type": 65535,                                                     
        "mask_guid": "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00",                                               
        "filter_priority": 0,                                             
        "filter_class": 30,                                                  
        "filter_type": 5,                                                   
        "filter_guid": "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00",
    }
  };

  try {
    console.log("\n* * * Waiting for SID...");
    const sid = await waitForSid();
    console.log(`SID       : ${sid}`);

    const authCrypto = encryptAuthCredentials(authConfig.username, authConfig.password, sid, authConfig.keyHex);
    const cmdauth = {
      "type": "cmd",
      "command": "auth",
      "args": {
         "iv": sid,
         "crypto": authCrypto
      }
    };

    console.log(`\n* * * Logging in as '${authConfig.username}'.`);
    console.log(`AUTH IV   : ${sid}`);
    console.log(`AUTH CRYPT: ${authCrypto}`);
    console.log(JSON.stringify(cmdauth, null, 2 ));
    await sendAndWait(cmdauth, 'auth');

    console.log("\n* * * Sending NOOP command.");
    console.log(JSON.stringify(cmdnoop, null, 2 ));
    await sendAndWait(cmdnoop, 'noop');

    console.log("\n* * * Sending VERSION command.");
    console.log(JSON.stringify(cmdver, null, 2 ));
    await sendAndWait(cmdver, 'version');

    console.log("\n* * * Sending COPYRIGHT command.");
    console.log(JSON.stringify(cmdcopy, null, 2 ));
    await sendAndWait(cmdcopy, 'copyright');

    console.log("\n* * * Open communication channel.");
    console.log(JSON.stringify(cmdopen, null, 2 ));
    await sendAndWait(cmdopen, 'open');

    console.log("\n* * * Set filter.");
    console.log(JSON.stringify(cmdFilter, null, 2 ));
    await sendAndWait(cmdFilter, 'setfilter');

    console.log("\n* * * Send event.");
    console.log(JSON.stringify(sendEvent, null, 2 ));
    ws.send(JSON.stringify(sendEvent));
  }
  catch (error) {
    console.error(`Command sequence failed: ${error.message}`);
    ws.close();
    return;
  }

  console.log("---------------------------------------------------------");
  console.log("Waiting for events (Abort with ctrl+c)")

  console.log("Will only receive CLASS1.CONTROL(30), TurnOn(5) now after filter");
  console.log("is set");
  console.log("---------------------------------------------------------");

});
  
ws.on('message', function incoming(data) {
  const message = toTextMessage(data);

  try {
    const reply = JSON.parse(message);
    printWs2Message(reply, message);

    if (pendingSidResolver) {
      const sid = extractSidFromReply(reply);
      if (sid) {
        pendingSidResolver.resolve(sid);
        return;
      }
    }

    if (pendingRequest) {
      if (reply.type === 'error') {
        pendingRequest.reject(new Error(`Error response: ${reply.code ?? ''} ${reply.message ?? ''}`.trim()));
        return;
      }

      if (commandMatches(reply, pendingRequest.expected)) {
        if (reply.code !== undefined && Number(reply.code) !== 0) {
          pendingRequest.reject(new Error(`Command failed (${reply.command}): code=${reply.code} message=${reply.message ?? ''}`.trim()));
        }
        else {
          pendingRequest.resolve(reply);
        }
      }
    }
  }
  catch (error) {
    console.log("\n* * * Non-JSON response:");
    console.log(message);
  }
});

File: firmware/test/ws/test_ws2.py
#!/usr/bin/env python

# VSCP ws2 client example  (Need python3)
# Demonstrates the use of the ws2 websocket interface of the VSCP daemon
# Sample event to send
#    send 0,20,3,,,0,-,15,14,13,12,11,10,9,8,7,6,5,4,3,2,0,0,1,35
# Original sample from: https://websockets.readthedocs.io/en/stable/intro.html
# Copyright 2020 Ake Hedman, the VSCP project - MIT license

from signal import signal, SIGINT
from sys import exit
import sys
import os
import asyncio
import pathlib
import ssl
import websockets
import base64
import json
import re

DEFAULT_URL = 'ws://localhost:8884/ws2'
DEFAULT_USERNAME = 'admin'
DEFAULT_PASSWORD = 'secret'
DEFAULT_KEY_HEX = 'A4A86F7D7E119BA3F0CD06881E371B98'

HEX_32_RE = re.compile(r'^[0-9a-fA-F]{32}$')


def get_aes_backend_name() -> str:
    try:
        from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
        _ = (Cipher, algorithms, modes)
        return 'cryptography'
    except ImportError:
        pass

    try:
        from Crypto.Cipher import AES
        _ = AES
        return 'pycryptodome'
    except ImportError:
        return 'missing'


def _pkcs7_pad(data: bytes, block_size: int = 16) -> bytes:
    pad_len = block_size - (len(data) % block_size)
    return data + bytes([pad_len]) * pad_len


def encrypt_auth_credentials(username: str, password: str, sid_hex: str, key_hex: str) -> str:
    if not HEX_32_RE.fullmatch(sid_hex or ''):
        raise ValueError(f"Invalid SID for IV: expected 32 hex chars, got '{sid_hex}'")

    if not HEX_32_RE.fullmatch(key_hex or ''):
        raise ValueError('Invalid AES-128 key: expected 32 hex chars')

    iv = bytes.fromhex(sid_hex)
    key = bytes.fromhex(key_hex)
    plaintext = _pkcs7_pad(f"{username};{password}".encode('utf-8'))

    try:
        from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
        encryptor = Cipher(algorithms.AES(key), modes.CBC(iv)).encryptor()
        ciphertext = encryptor.update(plaintext) + encryptor.finalize()
        return ciphertext.hex()
    except ImportError:
        try:
            from Crypto.Cipher import AES
            cipher = AES.new(key, AES.MODE_CBC, iv)
            return cipher.encrypt(plaintext).hex()
        except ImportError as ex:
            raise RuntimeError('AES dependency missing. Install `cryptography` or `pycryptodome`.') from ex


def find_sid_value(value):
    if isinstance(value, str) and HEX_32_RE.fullmatch(value):
        return value

    if isinstance(value, list):
        for item in value:
            sid = find_sid_value(item)
            if sid:
                return sid

    if isinstance(value, dict):
        for item in value.values():
            sid = find_sid_value(item)
            if sid:
                return sid

    return None

def get_cli_config():
    args = sys.argv[1:]

    if '--help' in args or '-h' in args:
        print('Usage: python test_ws2.py [url] [--url <url>] [--user <username>] [--password <password>] [--key <aes128-hex>]')
        print('Example: python test_ws2.py --url ws://192.168.1.50:8884/ws2 --user admin --password secret --key A4A86F7D7E119BA3F0CD06881E371B98')
        print('Env: WS2_URL, WS2_USERNAME, WS2_PASSWORD, WS2_KEY (fallback: VSCP_USERNAME, VSCP_PASSWORD, VSCP_KEY16)')
        exit(0)

    url = None
    username = None
    password = None
    key_hex = None
    positionals = []

    index = 0
    while index < len(args):
        arg = args[index]

        if arg == '--url' and index + 1 < len(args):
            url = args[index + 1]
            index += 2
            continue

        if arg.startswith('--url='):
            url = arg[len('--url='):]
            index += 1
            continue

        if arg == '--user' and index + 1 < len(args):
            username = args[index + 1]
            index += 2
            continue

        if arg.startswith('--user='):
            username = arg[len('--user='):]
            index += 1
            continue

        if arg == '--password' and index + 1 < len(args):
            password = args[index + 1]
            index += 2
            continue

        if arg.startswith('--password='):
            password = arg[len('--password='):]
            index += 1
            continue

        if arg == '--key' and index + 1 < len(args):
            key_hex = args[index + 1]
            index += 2
            continue

        if arg.startswith('--key='):
            key_hex = arg[len('--key='):]
            index += 1
            continue

        if not arg.startswith('-'):
            positionals.append(arg)

        index += 1

    if url is None and positionals:
        url = positionals[0]

    if not url:
        url = os.environ.get('WS2_URL', DEFAULT_URL)

    if not username:
        username = os.environ.get('WS2_USERNAME') or os.environ.get('VSCP_USERNAME') or DEFAULT_USERNAME

    if not password:
        password = os.environ.get('WS2_PASSWORD') or os.environ.get('VSCP_PASSWORD') or DEFAULT_PASSWORD

    if not key_hex:
        key_hex = os.environ.get('WS2_KEY') or os.environ.get('VSCP_KEY16') or DEFAULT_KEY_HEX

    return {
        'url': url,
        'username': username,
        'password': password,
        'key_hex': key_hex.strip(),
    }

def handler(signal_received, frame):
    print('SIGINT or CTRL-C detected. Exiting')
    exit(0)

async def connect(config):
    async with websockets.connect(
                 config['url'], ping_interval=20, ping_timeout=20, close_timeout=100) as websocket:

        # Get initial server response
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Initial response from server: {json.dumps(wsrply, indent=2)}")

        sid = find_sid_value(wsrply)
        if not sid:
            print('No SID in initial response, requesting challenge...')
            challenge = {
                "type": "cmd",
                "command": "challenge",
                "args": None
            }
            await websocket.send(json.dumps(challenge))
            response = await websocket.recv()
            wsrply = json.loads(response)
            print(f"Challenge response from server: {json.dumps(wsrply, indent=2)}")
            sid = find_sid_value(wsrply)

        if not sid:
            raise RuntimeError('No SID found in server response')

        auth_crypto = encrypt_auth_credentials(config['username'], config['password'], sid, config['key_hex'])

        # Log in as admin user

        cmdauth = {
            "type": "cmd",
            "command": "auth",
            "args": {
               "iv": sid,
               "crypto": auth_crypto
            }
        }

        print(f"\nLogging in as {config['username']}")
        print(f"AUTH IV   : {sid}")
        print(f"AUTH CRYPT: {auth_crypto}")
        await websocket.send(json.dumps(cmdauth))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        # NOOP

        cmdnoop = {
            "type": "cmd",
            "command": "noop",
            "args": None
        }

        print("\nNOOP command")
        await websocket.send(json.dumps(cmdnoop))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        # VERSION

        cmdver = {
            "type": "cmd",
            "command": "version",
            "args": None
        }

        print("\nVERSION command")
        await websocket.send(json.dumps(cmdver))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        # COPYRIGHT

        cmdcopy = {
            "type": "cmd",
            "command": "copyright",
            "args": None
        }

        print("\nCOPYRIGHT command")
        await websocket.send(json.dumps(cmdcopy))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        # Open Channel

        cmdopen = {
            "type": "cmd",
            "command": "open",
            "args": None
        }

        print(f"\nOpen channel  {json.dumps(cmdopen)}")
        await websocket.send(json.dumps(cmdopen))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")



        # Send event  - CLASS1.CONTROL, TurnOn
        sendEvent = {
            "type": "event",
            "event" : {
                "head" : 0,
                "obid": 0,
                "datetime": "2020-01-29T23:05:59Z",
                "timestamp": 0,
                "class": 30,
                "type": 5,
                "guid": "FF:FF:FF:FF:FF:FF:FF:F5:00:00:00:00:00:02:00:00",
                "data": [1,2,3,4,5,6]
            }
        }

        print(f"\nSend event  {json.dumps(sendEvent)}")
        await websocket.send(json.dumps(sendEvent))
        response = await websocket.recv()
        wsrply = json.loads(response)
        print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        # Set filter
        # Receive only CLASS1.CONTROL, TurnOn
        cmdFilter = {
            "type": "cmd",
            "command": "setfilter",
            "args": {
                "mask_priority": 0,
                "mask_class": 65535,
                "mask_type": 65535,                                                     
                "mask_guid": "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00",                                               
                "filter_priority": 0,                                             
                "filter_class": 30,                                                  
                "filter_type": 5,                                                   
                "filter_guid": "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00",
            }
        }

        # print(f"\nSet filter  {json.dumps(cmdFilter)}")
        # await websocket.send(json.dumps(cmdFilter))
        # response = await websocket.recv()
        # wsrply = json.loads(response)
        # print(f"Response from server: {json.dumps(wsrply, indent=2)}")

        print("Waiting for class=30, Type=5 as events (Active FILTER) (Abort with ctrl+c)")
        while True:
            response = await websocket.recv()
            wsrply = json.loads(response)
            print(f"Response from server: {json.dumps(wsrply, indent=2)}")
            e = wsrply["event"] # Get event
            # get without extracted event
            print(f"VSCP Class={wsrply['event']['vscpClass']}")
            # use extracted event 
            print(f"VSCP Type={e['vscpType']}")
            # Event can be edited and changed
            e['vscpType'] = 99
            print(f"Modified event: {json.dumps(e, indent=2)}")

if __name__ == '__main__':
    # Tell Python to run the handler() function when SIGINT is recieved
    signal(SIGINT, handler)
    cli_config = get_cli_config()
    backend = get_aes_backend_name()
    if backend == 'missing':
        print('AES backend : missing (install cryptography or pycryptodome)')
    else:
        print(f'AES backend : {backend}')
    print(f"Connecting to {cli_config['url']}")
    asyncio.get_event_loop().run_until_complete(connect(cli_config))
